### ISSUE TITLE
Runner: run multiple versions of each solver

### ISSUE THREAD

**Issue by siddharth-krishna** [open-energy-transition/solver-benchmark#13]
The goal here is for the website to have real data to show on the Solver Performance History page. We want to have benchmark data for the most recent solver version from each year (2023, 2022, ...) for each solver, and modify the graph to use this data to show how the runtime and memory consumption for each solver improves over time.

Some suggestions on how this can be achieved:

- [x] Find the highest solver version released in each of the past 5 years and create a `requirements-<year>.txt` file for each such year. (I think GLPK version cannot be managed by pip, so let's skip it for now, and just do this for HiGHS and SCIP.)
- [x] Write a bash wrapper script `benchmark_all.sh` that creates pip virtualenvs for each year's requirements file, activates each one in a loop, and calls `run_benchmarks.py`.
- [x] `run_benchmarks.py` should query the solver versions (within python, if possible, or else take it as argument from the bash script / take it from the requirements file) and additionally record `solver version` and `solver release year` in each line of the `benchmark_results.csv`
- [x] All pages of the website other than the Solver History page should only use the data for the latest version of each solver
- [x] The Solver History page should use the version and release year data to plot the graph appropriately

Edit: it looks like SCIP cannot also be managed purely through pip, so we will need another solution to manage older versions...
> For versions older than 4.4.0 installed via PyPI SCIP is not automatically installed. This means that SCIP must be installed yourself. If it is not installed globally, then the SCIPOPTDIR environment flag must be set, see [this page](https://pyscipopt.readthedocs.io/en/latest/build.html) for more details.

**Comment by jacek-oet** [comment]
>  Find the highest solver version released in each of the past 5 years and create a requirements-<year>.txt file for each such year. (I think GLPK version cannot be managed by pip, so let's skip it for now, and just do this for HiGHS and SCIP.)

@siddharth-krishna HiGHS solver only has versions available from 2022 onwards (as seen on PyPI). Should we skip running highspy for years before 2022?

**Comment by siddharth-krishna** [comment]
Sure, let's use whatever versions are available. 🙏

**Comment by jacek-oet** [comment]
@siddharth-krishna 

### 1. GLPK

```
conda search glpk
Loading channels: done
# Name                       Version           Build  Channel             
glpk                            4.59               0  conda-forge         
glpk                            4.60               0  conda-forge         
glpk                            4.61               0  conda-forge         
glpk                            4.61               1  conda-forge         
glpk                            4.62               0  conda-forge         
glpk                            4.63               0  conda-forge         
glpk                            4.65               0  conda-forge         
glpk                            4.65      h16a7912_1  conda-forge         
glpk                            4.65   h9202a9a_1003  conda-forge         
glpk                            4.65   h9202a9a_1004  conda-forge         
glpk                            4.65   ha14ba45_1001  conda-forge         
glpk                            4.65   he80fd80_1002  conda-forge         
glpk                             5.0      h445213a_0  conda-forge    `
```

https://lists.gnu.org/archive/cgi-bin/namazu.cgi?query=release+information&submit=Search&idxname=info-gnu

The latest release version of GLPK is 5.0 (released on Wed, 16 Dec 2020)     
we will use this version (5.0) for running benchmarks in 2020, 2021, 2022, 2023, and 2024.

### 2. SCIP

```
Running solver scip (version 7.0.3) on pypsa-gas+wind+sol+ely-ucgas-1-1h.lp (0)...
ERROR running solver. Captured output:
original problem has 87604 variables (26280 bin, 0 int, 0 impl, 61324 cont) and 157688 constraints
presolving:
   (0.5s) symmetry computation started: requiring (bin +, int -, cont +), (fixed: bin -, int +, cont -)
   (0.8s) no symmetry present
presolving (0 rounds: 0 fast, 0 medium, 0 exhaustive):
 0 deleted vars, 0 deleted constraints, 0 added constraints, 0 tightened bounds, 0 added holes, 0 changed sides, 0 changed coefficients
 0 implications, 0 cliques
presolved problem has 87604 variables (26280 bin, 0 int, 0 impl, 61324 cont) and 157688 constraints
 157688 constraints of type <linear>
Presolving Time: 0.54

 time | node  | left  |LP iter|LP it/n|mem/heur|mdpt |vars |cons |rows |cuts |sepa|confs|strbr|  dualbound   | primalbound  |  gap   | compl. 
p 184s|     1 |     0 | 58244 |     - |   locks|   0 |  87k|  65k|  65k|   0 |  0 |   0 |   0 | 0.000000e+00 | 8.466753e+10 |    Inf | unknown
* 297s|     1 |     0 |115854 |     - |    LP  |   0 |  87k|  65k|  65k|   0 |  0 |   0 |   0 | 8.466753e+10 | 8.466753e+10 |   0.00%| unknown
  297s|     1 |     0 |115854 |     - |   588M |   0 |  87k|  65k|  65k|   0 |  0 |   0 |   0 | 8.466753e+10 | 8.466753e+10 |   0.00%| unknown

SCIP Status        : problem is solved [optimal solution found]
Solving Time (sec) : 297.07
Solving Nodes      : 1
Primal Bound       : +8.46675266183077e+10 (2 solutions)
Dual Bound         : +8.46675266183077e+10
Gap                : 0.00 %

Traceback (most recent call last):
  File "/home/Documents/GitHub/oet-solver-benchmark/runner/run_solver.py", line 122, in <module>
    main(solver_name, input_file)
  File "/home/Documents/GitHub/oet-solver-benchmark/runner/run_solver.py", line 89, in main
    solver_result = solver.solve_problem(
  File "/home/miniforge3/envs/test-123/lib/python3.9/site-packages/linopy/solvers.py", line 292, in solve_problem
    return self.solve_problem_from_file(
  File "/home/miniforge3/envs/test-123/lib/python3.9/site-packages/linopy/solvers.py", line 1335, in solve_problem_from_file
    m.writeSol(m.getBestSol(), filename=path_to_string(solution_fn))
  File "src/pyscipopt/scip.pyx", line 4118, in pyscipopt.scip.Model.writeSol
OSError: [Errno 9] Bad file descriptor
Command exited with non-zero status 1
MaxResidentSetSizeKB=694016

Benchmarking complete.
```

We encountered an error

`pypsa-gas+wind+sol+ely-ucgas-1-1h,scip,ER,Error,,900,694.016,,,7.0.3,2021`


when running SCIP versions 7.0.3 and 7.0.1 in Linopy (latest releases for 2021 and 2020, respectively).

PySCIPOpt version: 3.5.0.

Should we skip these versions?

**Comment by siddharth-krishna** [comment]
Closed as completed by #55
